### PR TITLE
Save model with same filename

### LINKED ISSUE #1421: How to save model with same filename?
## 🚀 Feature

Hi @vfdev-5 ,

I found that `CheckpointHandler` can't use the fixed filename to save latest model:
https://github.com/pytorch/ignite/blob/v0.4.2/ignite/handlers/checkpoint.py#L149
But this is a very important requirement on our side...How can I achieve that?
For example, I want to always save the best model as `model.pt` after validations during training.

Thanks.

### PR DESCRIPTION
**Pull request by sdesrozis**
Fixes #1421 

Description:

This solves the problem to do not save model with an existing filename. Now, filename is overwritten.

Check list:
* [ ] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/handlers/checkpoint.py [hunk 1]
```diff
@@ -147,8 +147,8 @@ class Checkpoint(Serializable):
         ``30000-checkpoint-94.pt``
 
         **Warning:** Please, keep in mind that if filename collide with already used one to saved a checkpoint,
-        new checkpoint will not be stored. This means that filename like ``checkpoint.pt`` will be saved only once
-        and will not be overwritten by newer checkpoints.
+        new checkpoint will replace the older one. This means that filename like ``checkpoint.pt`` will be saved
+        every call and will always be overwrittDiskSaveren by newer checkpoints.
```
Comment (vfdev-5) [inline ignite/handlers/checkpoint.py:151]:
```suggestion
        every call and will always be overwritten by newer checkpoints.
```

Comment (sdesrozis) [reply ignite/handlers/checkpoint.py]: Sorry 😅

File: ignite/handlers/checkpoint.py [hunk 2]
```diff
@@ -350,19 +350,24 @@ def __call__(self, engine: Engine) -> None:
...
-                item = self._saved.pop(0)
+            filename_already_exists = any(item.filename == filename for item in self._saved)
+
+            if filename_already_exists:
                 if isinstance(self.save_handler, BaseSaveHandler):
-                    self.save_handler.remove(item.filename)
+                    self.save_handler.remove(filename)
+                # list is purged
+                self._saved = [item for item in self._saved if item.filename != filename]
+            else:
```
Comment (vfdev-5) [inline ignite/handlers/checkpoint.py:366]: We can use `elif` instead of `else -> if`

File: ignite/handlers/checkpoint.py [hunk 3]
```diff
@@ -350,19 +350,24 @@ def __call__(self, engine: Engine) -> None:
...
-            if not self._check_lt_n_saved():
-                item = self._saved.pop(0)
+            filename_already_exists = any(item.filename == filename for item in self._saved)
+
+            if filename_already_exists:
                 if isinstance(self.save_handler, BaseSaveHandler):
-                    self.save_handler.remove(item.filename)
+                    self.save_handler.remove(filename)
+                # list is purged
+                self._saved = [item for item in self._saved if item.filename != filename]
```
Comment (vfdev-5) [inline ignite/handlers/checkpoint.py:365]: I'd like to simplify `any(item.filename == filename for item in self._saved)` and `self._saved = [item for item in self._saved if item.filename != filename]` somehow.

Comment (sdesrozis) [reply ignite/handlers/checkpoint.py:365]: Ok I think about it.

File: ignite/handlers/checkpoint.py [hunk 4]
```diff
@@ -356,13 +356,12 @@ def __call__(self, engine: Engine) -> None:
...
-            filename_already_exists = any(item.filename == filename for item in self._saved)
+            saved = [item for item in self._saved if item.filename != filename]
 
-            if filename_already_exists:
+            if self._saved != saved:
                 if isinstance(self.save_handler, BaseSaveHandler):
                     self.save_handler.remove(filename)
-                # list is purged
-                self._saved = [item for item in self._saved if item.filename != filename]
+                self._saved = saved
```
Comment (vfdev-5) [inline ignite/handlers/checkpoint.py:364]: Thanks Sylvain! I'm a bit bothered by the fact that we redefine `self._saved` every time if `filename_already_exists`. I was thinking if we could get the Item that has existing filename and then call remove on the list of `self._saved` ? What do you think ?

Comment (sdesrozis) [reply ignite/handlers/checkpoint.py:364]: I agree with you.

File: tests/ignite/handlers/test_checkpoint.py [hunk 5]
```diff
@@ -1446,3 +1449,23 @@ def test_checkpoint_load_state_dict():
     sd = {"saved": [(0, "model_0.pt"), (10, "model_10.pt"), (20, "model_20.pt")]}
     checkpointer.load_state_dict(sd)
     assert checkpointer._saved == true_checkpointer._saved
+
+
+def test_checkpoint_fixed_filename():
+    save_handler = MagicMock(spec=BaseSaveHandler)
+    model = DummyModel()
+    to_save = {"model": model}
+    checkpointer = Checkpoint(to_save, save_handler=save_handler, n_saved=None, filename_pattern="{name}.{ext}")
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_checkpoint.py:1458]: Could you please test with different values of `n_saved`: None and 1 or 2
and call `checkpointer(trainer)` at least 5-10 times.

**Review by vfdev-5 — APPROVED** [review]
Thanks a lot, Sylvain ! I like your last solution 👍 
I left a nit comment to improve a bit the test.

**Pull request merged**